### PR TITLE
feat: add endpoints to manage user highlight reactions

### PR DESCRIPTION
**Pull request by brandonroberts**
## What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Description

- Adds endpoints to add, remove user reactions for a highlight
- Adds endpoints to retrieve overall and user-specific reactions for a highlight

## Related Tickets & Documents

Closes https://github.com/open-sauced/ops-backend/issues/15

## Mobile & Desktop Screenshots/Recordings

image

image

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: migrations/20230405125001_add_emojis.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+create table if not exists public.emojis
+(
+  id bigint not null generated by default as identity ( increment 1 start 1 minvalue 1 maxvalue 9223372036854775807 cache 1 ),
```
Comment (0-vortex) [inline migrations/20230405125001_add_emojis.sql:3]: https://docs.github.com/en/rest/emojis?apiVersion=2022-11-28#get-emojis don't have ids, can we move this over to uuids?

File: migrations/20230405125001_add_user_highlight_reactions.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
+create table if not exists public.user_highlight_reactions
+(
+  id bigint not null generated by default as identity ( increment 1 start 1 minvalue 1 maxvalue 9223372036854775807 cache 1 ),
```
Comment (0-vortex) [inline migrations/20230405125001_add_user_highlight_reactions.sql:3]: same for uuids here

File: src/emoji/emoji.controller.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+@ApiTags("Emojis service")
+export class EmojiController {
+  constructor (
+    private emojiService: EmojiService,
+  ) {}
+
+  @Get("/")
+  @ApiOperation({
+    operationId: "findAllInsightsByUserId",
+    summary: "Listing all insights for a user and paginate them",
```
Comment (0-vortex) [inline src/emoji/emoji.controller.ts:20]: wrong description here

File: src/user/entities/user-highlight-reaction.entity.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  })
+  @PrimaryGeneratedColumn()
+  public id!: number;
+
+  @ApiModelProperty({
+    description: "Highlight identifier",
+    example: 71359796,
+  })
+  @Column({ type: "integer" })
+  public highlight_id!: number;
```
Comment (0-vortex) [inline src/user/entities/user-highlight-reaction.entity.ts:22]: this needs the relationship added to both entities

File: src/user/entities/user-highlight-reaction.entity.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  @ApiModelProperty({
+    description: "User identifier",
+    example: 237133,
+  })
+  @Column({
+    type: "integer",
+    select: false,
+  })
+  public user_id!: number;
```
Comment (0-vortex) [inline src/user/entities/user-highlight-reaction.entity.ts:32]: this needs the relationship added to both entities

File: src/user/user-highlight.controller.ts [hunk 6]
```diff
@@ -86,4 +87,65 @@ export class UserHighlightsController {
...
+  })
+  @ApiBearerAuth()
+  @UseGuards(SupabaseGuard)
+  @ApiNotFoundResponse({ description: "Unable to delete user highlight reaction" })
+  @ApiBadRequestResponse({ description: "Invalid request" })
+  async deleteHighlightReactionForUser (
+    @UserId() userId: number,
+      @Param("id") highlightId: number,
+      @Param("emojiId") emojiId: number,
+  ): Promise<void> {
```
Comment (0-vortex) [inline src/user/user-highlight.controller.ts:146]: is this empty 200 response?

File: src/user/user-highlights.service.ts [hunk 7]
```diff
@@ -140,4 +143,74 @@ export class UserHighlightsService {
...
+      if (!reactionExists.deleted_at) {
+        throw new ConflictException("You have already added this reaction for this highlight");
+      }
+
+      await this.userHighlightReactionRepository.restore(reactionExists.id);
+
+      return reactionExists;
+    }
+
+    return this.userHighlightReactionRepository.save({
```
Comment (0-vortex) [inline src/user/user-highlights.service.ts:206]: save can be finicky, we should use normal array inserts and handle the returns

**Review by 0-vortex — CHANGES_REQUESTED** [review]
Needs small polishes, the `.save` would argue is optional tho ❤️ 🍕

File: src/user/user-highlight.controller.ts (hunk 6 above)
Comment (brandonroberts) [reply src/user/user-highlight.controller.ts:146]: Yes

**Pull request merged**